rocAL/rocAL/source/bounding_box_graph.cpp

```diff
@@ -147,3 +147,159 @@ void BoundingBoxGraph::update_random_bbox_meta_data(MetaDataBatch *input_meta_da
+{
+    #pragma omp parallel for 
+    for (int i = 0; i < full_batch_meta_data->size(); i++)
+    {
+        auto bb_count = full_batch_meta_data->get_bb_labels_batch()[i].size();
+        int labels_buf[bb_count];
+        float coords_buf[bb_count * 4];
+        memcpy(labels_buf, full_batch_meta_data->get_bb_labels_batch()[i].data(), sizeof(int) * bb_count);
+        memcpy(coords_buf, full_batch_meta_data->get_bb_cords_batch()[i].data(), full_batch_meta_data->get_bb_cords_batch()[i].size() * sizeof(BoundingBoxCord));
+        BoundingBoxCords encoded_bb;
+        BoundingBoxLabels encoded_labels;
+        BoundingBoxCords bb_coords;
+        BoundingBoxLabels bb_labels;
+        unsigned anchors_size = anchors.size() / 4; // divide the anchors_size by 4 to get the total number of anchors
+        //Calculate Ious
+        //ious size - bboxes count x anchors count
+        std::vector<float> ious(bb_count * anchors_size);
+        bb_coords.resize(bb_count);
+        bb_labels.resize(bb_count);
+        encoded_bb.resize(anchors_size);
+        encoded_labels.resize(anchors_size);
+        for (uint bb_idx = 0; bb_idx < bb_count; bb_idx++)
+        {
+            int m = bb_idx * 4;
+            BoundingBoxCord box;
+            box.l = coords_buf[m];
+            box.t = coords_buf[m + 1];
+            box.r = coords_buf[m + 2];
+            box.b = coords_buf[m + 3];
+            auto iou_rows = ious.data() + (bb_idx * (anchors_size));
+            calculate_ious_for_box(iou_rows, box, anchors);
+            bb_coords[bb_idx] = box;
+            bb_labels[bb_idx] = labels_buf[bb_idx];
+        }
+        
+        // Depending on the matches ->place the best bbox instead of the corresponding anchor_idx in anchor
+        for (unsigned anchor_idx = 0; anchor_idx < anchors_size; anchor_idx++)
+        {
+            int m = anchor_idx * 4;
+            BoundingBoxCord box_bestidx, _anchor, _anchor_xcyxwh;
+            _anchor.l = anchors[m];
+            _anchor.t = anchors[m + 1];
+            _anchor.r = anchors[m + 2];
+            _anchor.b = anchors[m + 3];
+            const auto best_idx = find_best_box_for_anchor(anchor_idx, ious, bb_count, anchors_size);
+            // Filter matches by criteria
+            if (ious[(best_idx * anchors_size) + anchor_idx] > criteria) //Its a match
+            {
+                //YTD: Need to add a new structure for xc,yc,w,h similar to l,t,r,b as a part of metadata
+                    box_bestidx.l = 0.5 * (bb_coords.at(best_idx).l + bb_coords.at(best_idx).r); //xc
+                    box_bestidx.t = 0.5 * (bb_coords.at(best_idx).t + bb_coords.at(best_idx).b); //yc
+                    box_bestidx.r = (-bb_coords.at(best_idx).l + bb_coords.at(best_idx).r);      //w
+                    box_bestidx.b = (-bb_coords.at(best_idx).t + bb_coords.at(best_idx).b);      //h
+
+                if (offset)
+                {
+                    box_bestidx.l *= scale; //xc
```

Comment:
Please add comment on what the math is doing and reference if any

Reply:
Done